torch/_inductor/runtime/triton_bundler.py

```diff
@@ -0,0 +1,127 @@
+import dataclasses
+import logging
+import os
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import torch
+from torch._dynamo.utils import counters
+
+from .runtime_utils import cache_dir
+
+
+log = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonBundleEntry:
+    kernel_hash: str
+    device: int
+    directory: str
+
+
+@dataclasses.dataclass
+class TritonKernelArtifact:
+    filename: str
+    payload: bytes = dataclasses.field(repr=False)  # Do not display binary
+
+
+@dataclasses.dataclass
+class TritonKernelArtifacts:
+    kernel_hash: str
+    device: int
+    artifacts: List[TritonKernelArtifact]
+
+
+class TritonBundler:
+    # It is safe for multiple threads to insert entries as insert operation
+    # of dict is atomic
+    _entries: Optional[Dict[TritonBundleEntry, None]] = None
+
+    @staticmethod
+    def is_enabled() -> bool:
+        from torch._inductor import config
+
+        if (b := config.bundle_triton_into_fx_graph_cache) is not None:
+            return b
+
+        if not config.is_fbcode():
+            return False
+
+        return torch._utils_internal.justknobs_check(
+            "pytorch/remote_cache:bundle_triton_into_fx_graph_cache"
+        )
+
+    @classmethod
+    def begin_compile(cls) -> None:
+        if not TritonBundler.is_enabled():
+            return
+        assert cls._entries is None
+        cls._entries = {}
+
+    @classmethod
+    def put(cls, kernel_hash: str, device: int) -> None:
+        if (entries := cls._entries) is not None:
+            # CachingAutotuner.__init__ unconditionally sets TRITON_CACHE_DIR
+            triton_cache_dir = os.getenv("TRITON_CACHE_DIR")
+            assert triton_cache_dir is not None
+            entries[TritonBundleEntry(kernel_hash, device, triton_cache_dir)] = None
+
+    @classmethod
+    def collect(cls) -> List[TritonKernelArtifacts]:
+        if not TritonBundler.is_enabled():
+            cls._entries = None
```

Comment:
I am a bit confused about thread safety here. You are talking about `_entries` writes OK from multiple threads to be atomic. Is collect a sync point then or something?

Reply:
I'm probably being overly cautious here. collect is executed when we are writing the inductor cache entry to disk/remote, so at this point we expect everything to have completed and this is true by design (no need for any sync primitive).